monai/data/image_writer.py

```diff
@@ -0,0 +1,395 @@
+        - data type of the output image
+            - as part of ``resample_if_needed()``
+
+    Subclasses of this class should implement the backend-specific functions:
+
+        - ``set_data_array()`` to set the data array (input must be numpy array or torch tensor)
+            - this method sets the backend object's data part
+        - ``set_metadata()`` to set the metadata and output affine
+            - this method sets the metadata including affine handling and image resampling
+        - backend-specific data object
+            - ``create_backend_obj()``
+        - backend-specific writing function
+            - ``write()``
+
+    The primary usage is:
+
+    .. code-block:: python
+
+        writer = MyWriter()  # subclass of ImageWriter
+        writer.set_data_array(data_array)
+        writer.set_metadata(meta_dict).write(filename_or_obj)
+        writer.write(filename)
+
+    Create an image writer object based on ``data_array`` and ``metadata``.
+    Spatially it supports up to three dimensions (with the resampling step
+    supports both 2D and 3D).
+
+    When saving multiple time steps or multiple channels `data_array`, time
+    and/or modality axes should be the at the `channel_dim`. For example,
+    the shape of a 2D eight-class and channel_dim=0, the segmentation
+    probabilities to be saved could be `(8, 64, 64)`; in this case
+    ``data_array`` will be converted to `(64, 64, 1, 8)` (the third
+    dimension is reserved as a spatial dimension).
+
+    The ``metadata`` could optionally have the following keys:
+
+        - ``'original_affine'``: for data original affine, it will be the
+            affine of the output object, defaulting to an identity matrix.
+        - ``'affine'``: it should specify the current data affine, defaulting to an identity matrix.
+        - ``'spatial_shape'``: for data output spatial shape.
+
+    When ``metadata`` is specified and ``resample=True``, the saver will
+    try to resample data from the space defined by `"affine"` to the space
+    defined by `"original_affine"`, for more details, please refer to the
+    ``resample_if_needed`` method.
+    """
+
+    def __init__(self, **kwargs):
+        """the constructor supports adding new instance members."""
+        self.data_obj = None
+        for k, v in kwargs.items():
+            setattr(self, k, v)
+
+    def set_data_array(self, data_array, **kwargs):
+        raise NotImplementedError(f"Subclasses of {self.__class__.__name__} must implement this method.")
+
+    def set_metadata(self, meta_dict: Optional[Mapping], **options):
+        raise NotImplementedError(f"Subclasses of {self.__class__.__name__} must implement this method.")
+
+    def write(self, filename_or_obj: PathLike, verbose: bool = True, **kwargs):
+        """subclass should implement this method to call the backend-specific writing APIs."""
+        if verbose:
+            logger.info(f"writing: {filename_or_obj}")
+
+    @classmethod
```

Comment:
May I know why using `classmethod` here? Maybe making it into a utility function directly?

Thanks.

Reply:
subclasses can extend this method for different backend, and they share some common logic such as casting torch tensor to ndarray.

Reply:
OK, I see. Seems you don't want to implement it as `static` function and it also works with other class functions.

Thanks.